openhands/controller/agent_controller.py

```diff
@@ -223,6 +223,21 @@ async def on_event(self, event: Event):
 
         # if the event is not filtered out, add it to the history
         if not any(isinstance(event, filter_type) for filter_type in self.filter_out):
+            # Check if adding this event would exceed context window
+            if self.agent.llm.config.max_input_tokens is not None:
+                # Create temporary history with new event
+                temp_history = self.state.history + [event]
+                try:
+                    token_count = self.agent.llm.get_token_count(temp_history)
```

Comment:
I'm afraid this line doesn't really do what we want. We need to do the token counting on the messages that are sent to the LLM API. OpenHands works with events, and these are events, which it then 'translates' into messages and sends to the LLM API. You can see, as you try to make it work on your machine for your use case, that the number of tokens will not match, some events differ more than others but they all differ. 😅 

An alternative is to define a custom exception like, let's say, TokenLimitExceeded, [here](https://github.com/All-Hands-AI/OpenHands/blob/main/openhands/core/exceptions.py), and move this check to the [LLM class](https://github.com/All-Hands-AI/OpenHands/blob/83add629917de56f4f7a238df36977b6a88bbcfd/openhands/llm/llm.py#L190), then raise the exception when the token comparison fails. Then maybe the exception can be treated like this [ContextWindowExceededError](https://github.com/All-Hands-AI/OpenHands/blob/83add629917de56f4f7a238df36977b6a88bbcfd/openhands/controller/agent_controller.py#L502) is. What do you think, does that make sense?

Reply:
ahh I see I think I understand.  I was getting unpredictable behavior last night and that makes more sense and this definitely doesn't work.  I'll take another crack at it when I get a chance today.  

I do have a dumb question but I can't recreate it predictably.  How is the config.toml file loaded?  Is it only with certain commands? I can't tell if I broke something on my machine or I am assuming incorrect behavior

Reply:
Let's figure it out! There is something a bit unpredictable about it: when you set values there, then run either with UI or with `main.py` or with `cli.py`, then config.toml is loaded. However, if running with UI, any settings that are defined in the UI (in the Settings window) will _override_ the toml values.
Could that be what happens?

Reply:
And it's absolutely a good question! We're still trying to document it properly, the simple thing that it turns out a bit difficult does say that something's wrong and we should rethink some of it.
For reference: https://github.com/All-Hands-AI/OpenHands/issues/3220